lib/src/gestures/flutter_map_state_controller.dart

```diff
@@ -0,0 +1,440 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter_map/plugin_api.dart';
+import 'package:flutter_map/src/gestures/flutter_map_interactive_viewer.dart';
+import 'package:flutter_map/src/map/flutter_map_state_controller_interface.dart';
+import 'package:flutter_map/src/map/map_controller_impl.dart';
+import 'package:latlong2/latlong.dart';
+
+// This controller is for internal use. All updates to the state should be done
+// by calling methods of this class to ensure consistency.
+class FlutterMapStateController extends ValueNotifier<FlutterMapState>
+    implements FlutterMapStateControllerInterface {
+  late final FlutterMapInteractiveViewerState _interactiveViewerState;
+  late MapControllerImpl _mapControllerImpl;
+
+  FlutterMapStateController(MapOptions options)
+      : super(FlutterMapState.initialState(options));
+
+  // Link the viewer state with the controller. This should be done once when
+  // the FlutterMapInteractiveViewerState is initialized.
+  set interactiveViewerState(
+    FlutterMapInteractiveViewerState interactiveViewerState,
+  ) =>
+      _interactiveViewerState = interactiveViewerState;
+
+  void linkMapController(MapControllerImpl mapControllerImpl) {
+    _mapControllerImpl = mapControllerImpl;
+    _mapControllerImpl.stateController = this;
+  }
```

Comment:
Should this be a setter?

Reply:
Given that it's not just setting the mapControllerImpl, it's also setting the MapControllerImpl's FlutterMapStateController, I think it's clearer to keep it a method with 'link' in the name.

Reply:
Sure, guess that makes sense.